File: app/static/js/cards.js
/* Visuel de carte canvas (§13) — mi-rappette du rendu serveur,
   mais calqué pour les exports (story, carré, paysage). */

const SIZES = {
  story: [1080, 1920],
  square: [1080, 1080],
  landscape: [1200, 630],
};
const MAX_CHARS = { story: 1400, square: 900, landscape: 420 };
const C = {
  ink: "#14110d",
  panel: "#1c1712",
  ivory: "#f0e6d0",
  gold: "#b08d3e",
  goldSoft: "#c8a663",
  muted: "#9a8767",
};

function spacingText(ctx, text, font, color, cx, cy, spacing) {
  ctx.save();
  ctx.font = font;
  ctx.fillStyle = color;
  ctx.textBaseline = "alphabetic";
  let total = 0;
  for (const ch of text) total += ctx.measureText(ch).width + spacing;
  let x = cx - (total - spacing) / 2;
  for (const ch of text) {
    ctx.fillText(ch, x, cy);
    x += ctx.measureText(ch).width + spacing;
  }
  ctx.restore();
}

function ornamentRow(ctx, cx, cy, halfLen, color) {
  ctx.save();
  ctx.strokeStyle = color;
  ctx.lineWidth = 1;
  ctx.beginPath();
  ctx.moveTo(cx - halfLen, cy);
  ctx.lineTo(cx - 16, cy);
  ctx.moveTo(cx + 16, cy);
  ctx.lineTo(cx + halfLen, cy);
  ctx.stroke();
  ctx.beginPath();
  const d = 6;
  ctx.moveTo(cx, cy - d);
  ctx.lineTo(cx + d, cy);
  ctx.lineTo(cx, cy + d);
  ctx.lineTo(cx - d, cy);
  ctx.closePath();
  ctx.stroke();
  ctx.restore();
}

function wrap(ctx, text, maxWidth) {
  const words = text.split(/\s+/);
  const lines = [];
  let cur = "";
  for (const w of words) {
    const trial = cur ? cur + " " + w : w;
    if (ctx.measureText(trial).width <= maxWidth) cur = trial;
    else {
      if (cur) lines.push(cur);
      cur = w;
    }
  }
  if (cur) lines.push(cur);
  return lines;
}

export function renderCard(canvas, entry, format = "story") {
  const [W, H] = SIZES[format] || SIZES.story;
  canvas.width = W;
  canvas.height = H;
  const ctx = canvas.getContext("2d");
  const s = W / 1200;

  ctx.fillStyle = C.ink;
  ctx.fillRect(0, 0, W, H);

  // vignette douce
  const g = ctx.createRadialGradient(W / 2, H / 3, H / 8, W / 2, H / 3, H);
  g.addColorStop(0, "rgba(60,45,20,0.5)");
  g.addColorStop(1, "rgba(0,0,0,0.3)");
  ctx.fillStyle = g;
  ctx.fillRect(0, 0, W, H);

  const margin = 45 * s;
  ctx.strokeStyle = C.gold;
  ctx.lineWidth = 3;
  ctx.strokeRect(margin, margin, W - 2 * margin, H - 2 * margin);
  ctx.lineWidth = 1;
  ctx.strokeRect(margin + 12, margin + 12, W - 2 * margin - 24, H - 2 * margin - 24);

  const innerW = W - 2 * margin - 60;
  ctx.textBaseline = "alphabetic";
  let y = margin + 46 * s;

  spacingText(ctx, "LE GRAND PHILOSOPHE", `700 ${26 * s}px Fraunces, Georgia, serif`,
              C.gold, W / 2, y, 6 * s);
  y += 46 * s;
  ornamentRow(ctx, W / 2, y, innerW * 0.3, C.gold);
  y += 46 * s;

  // corps avec capitale ornée
  let text = entry.text || "";
  const limit = MAX_CHARS[format] || MAX_CHARS.story;
  if (text.length > limit) {
    text = text.slice(0, limit).replace(/\s+\S*$/, "") + "…";
  }
  text = text.replace(/^#+\s*/m, "").replace(/\n+/g, " ").trim();

  const bodySize = 30 * s;
  const lineH = bodySize * 1.55;
  ctx.font = `${bodySize}px "EB Garamond", Georgia, serif`;

  const capSize = bodySize * 2.6;
  const first = text[0] || "";
  const rest = text.slice(1);
  ctx.font = `500 ${capSize}px Fraunces, Georgia, serif`;
  ctx.fillStyle = C.goldSoft;
  ctx.fillText(first, margin + 30, y + capSize * 0.45);
  const capW = ctx.measureText(first).width + 14 * s;

  ctx.font = `${bodySize}px "EB Garamond", Georgia, serif`;
  ctx.fillStyle = C.ivory;
  const tx = margin + 30 + capW;
  const attrH = entry.attribution ? 150 * s : 0;
  const lines = wrap(ctx, rest, innerW - capW);
  const maxLines = Math.max(1, Math.floor((H - y - margin - 60 * s - attrH) / lineH));
  // Sur le format story, le bloc de texte est centré verticalement.
  if (format === "story") {
    const totalH = Math.min(lines.length, maxLines) * lineH + capSize;
    y = Math.max(y, (H - totalH - margin * 2 - attrH) / 2);
  }
  for (const line of lines.slice(0, maxLines)) {
    ctx.fillText(line, tx, y + bodySize);
    y += lineH;
  }

  if (entry.attribution) {
    y = H - margin - 150 * s;
    ctx.font = `italic ${bodySize * 0.85}px "EB Garamond", Georgia, serif`;
    ctx.fillStyle = C.muted;
    ctx.textAlign = "right";
    const attrLines = wrap(ctx, entry.attribution, innerW).slice(0, 2);
    let ay = y;
    for (const line of attrLines) {
      ctx.fillText(line, W - margin - 30, ay);
      ay += bodySize;
    }
    ctx.textAlign = "left";
    // badge anti-confusion (§9.2)
    const badge = "citation humoristique · attribution fictive";
    ctx.font = `${13 * s}px Inter, system-ui, sans-serif`;
    const bw = ctx.measureText(badge).width + 28 * s;
    const bx = W / 2 - bw / 2;
    const by = H - margin - 78 * s;
    ctx.strokeStyle = C.gold;
    ctx.lineWidth = 1;
    if (ctx.roundRect) {
      ctx.beginPath();
      ctx.roundRect(bx, by, bx + bw, by + 30 * s, 14 * s);
      ctx.stroke();
    } else {
      ctx.strokeRect(bx, by, bw, 30 * s);
    }
    ctx.fillStyle = C.gold;
    ctx.fillText(badge, bx + 14 * s, by + 20 * s);
  }

  const fy = H - margin - 46 * s;
  ornamentRow(ctx, W / 2, fy, innerW * 0.25, C.gold);
  spacingText(ctx, "LEGRANDPHILOSOPHE", `${13 * s}px Inter, sans-serif`, C.muted, W / 2, fy + 18 * s, 4 * s);
}

export function downloadCard(canvas, filename = "grande-reflexion.png") {
  canvas.toBlob((blob) => {
    if (!blob) return;
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    a.click();
    URL.revokeObjectURL(url);
  }, "image/png");
}

export async function copyCard(canvas) {
  const blob = await new Promise((r) => canvas.toBlob(r, "image/png"));
  if (!blob) throw new Error("canvas vide");
  await navigator.clipboard.write([new ClipboardItem({ "image/png": blob })]);
}

export async function shareCard(canvas, title) {
  const blob = await new Promise((r) => canvas.toBlob(r, "image/png"));
  if (navigator.canShare?.({ files: [new File([blob], "reflexion.png", { type: "image/png" })] })) {
    await navigator.share({ title, files: [new File([blob], "reflexion.png", { type: "image/png" })] });
  } else {
    await shareFallback(title);
  }
}

async function shareFallback(title) {
  const text = window.location.href;
  if (navigator.share) await navigator.share({ title, text });
}

export async function publishCard(entry) {
  const res = await fetch("/api/cards", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      input_text: entry.input,
      generated_text: entry.text,
      attribution: entry.attribution,
      mode: entry.mode,
      styles: entry.params?.styles || [],
    }),
  });
  if (!res.ok) throw new Error("publication impossible");
  const data = await res.json();
  return data.path; // /carte/<id>
}

File: app/cards.py
"""Magasin de cartes publiques + rendu PNG serveur (Pillow).

Chaque carte partagée possède une URL unique ``/carte/<id>`` servie avec
des métadonnées Open Graph dynamiques ; l'image OG elle-même provient du
rendu serveur ``/carte/<id>.png``.

Le magasin est volontairement simple : un fichier JSON local. Une base
SQL serait le pas naturel si le produit passait freemium (§12).
"""

from __future__ import annotations

import json
import os
import secrets
import threading

from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter, ImageFont

DATA_DIR = Path(os.environ.get("GRAND_DATA_DIR", "data"))
CARDS_FILE = DATA_DIR / "cards.json"

FONT_SERIF = "/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf"
FONT_SERIF_BOLD = "/usr/share/fonts/truetype/dejavu/DejaVuSerif-Bold.ttf"
FONT_SANS = "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf"

CARD_SIZES = {
    "story": (1080, 1920),
    "square": (1080, 1080),
    "landscape": (1200, 630),
}
MAX_CHARS = {"story": 1400, "square": 900, "landscape": 420}

# Palette(s) templating
PALETTE = {
    "ink": (20, 17, 12),
    "gold": (176, 141, 62),
    "gold_soft": (200, 166, 99),
    "ivory": (241, 232, 214),
    "muted": (146, 131, 104),
}

_lock = threading.Lock()

# Le fichier JSON ne doit pas grossir sans borne : on écarte les plus
# anciennes entrées au-delà d'un plafond (FIFO).
MAX_CARDS = 5000


def _load() -> dict:
    if CARDS_FILE.exists():
        try:
            return json.loads(CARDS_FILE.read_text(encoding="utf-8"))
        except Exception:
            return {}
    return {}


def _save(store: dict) -> None:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    tmp = CARDS_FILE.with_suffix(".tmp")
    tmp.write_text(json.dumps(store, ensure_ascii=False), encoding="utf-8")
    os.replace(tmp, CARDS_FILE)


def _evict_oldest(store: dict) -> None:
    while len(store) > MAX_CARDS:
        # dict ordonné => la première clé est la plus ancienne créée
        del store[next(iter(store))]



def create_card(payload: dict) -> str:
    with _lock:
        store = _load()
        for _ in range(8):
            cid = secrets.token_hex(5)
            if cid not in store:
                store[cid] = payload
                _evict_oldest(store)
                _save(store)
                return cid
    raise RuntimeError("impossible d'allouer un identifiant de carte")


def get_card(card_id: str) -> dict | None:
    with _lock:
        return _load().get(card_id)


def _font(path: str, size: int) -> ImageFont.FreeTypeFont:
    return ImageFont.truetype(path, size)


def _spaced(draw: ImageDraw.ImageDraw, x: float, y: float, text: str,
            font: ImageFont.FreeTypeFont, fill, tracking: int = 6) -> float:
    """Dessine un texte avec approches (letter-spacing) centrées."""
    total = sum(draw.textlength(ch, font=font) + tracking for ch in text) - tracking
    cx = x - total / 2
    for ch in text:
        w = draw.textlength(ch, font=font)
        draw.text((cx, y), ch, font=font, fill=fill)
        cx += w + tracking
    return total


def _wrap(draw: ImageDraw.ImageDraw, text: str,
          font: ImageFont.FreeTypeFont, max_width: float) -> list[str]:
    words = text.split()
    lines: list[str] = []
    current = ""
    for word in words:
        trial = f"{current} {word}".strip()
        if draw.textlength(trial, font=font) <= max_width:
            current = trial
        else:
            if current:
                lines.append(current)
            current = word
    if current:
        lines.append(current)
    return lines


def _ornament_row(draw: ImageDraw.ImageDraw, cx: float, y: float, width: float,
                  color) -> None:
    """Ligne fine avec un losange au centre — le sceau du produit."""
    half = width / 2
    draw.line([(cx - half, y), (cx - 16, y)], fill=color, width=1)
    draw.line([(cx + 16, y), (cx + half, y)], fill=color, width=1)
    d = 7
    draw.polygon([(cx, y - d), (cx + d, y), (cx, y + d), (cx - d, y)],
                 outline=color, width=1)


def render_card(card: dict, size_key: str = "landscape") -> Image.Image:
    """Rend le visuel de la carte : encre, or, capitale, filigrane."""
    if size_key not in CARD_SIZES:
        size_key = "landscape"
    W, H = CARD_SIZES[size_key]
    img = Image.new("RGB", (W, H), PALETTE["ink"])
    draw = ImageDraw.Draw(img)

    # Vignette douce (façon lampe sur papier)
    small = Image.radial_gradient("L").resize((W, H)).filter(ImageFilter.GaussianBlur(26))
    glow = Image.new("RGB", (W, H), (36, 29, 19))
    img.paste(glow, mask=small.point(lambda v: int(v * 0.30)))
    draw = ImageDraw.Draw(img)

    margin = 45 if W > 700 else 34
    # double cadre doré
    draw.rectangle([margin, margin, W - margin, H - margin], outline=PALETTE["gold"], width=3)
    draw.rectangle([margin + 12, margin + 12, W - margin - 12, H - margin - 12],
                   outline=PALETTE["gold"], width=1)

    inner_w = W - 2 * margin - 60
    y = margin + 40

    # en-tête : nom du produit, lettres espacées
    head_size = max(20, int(26 * W / 1200))
    _spaced(draw, W / 2, y, "LE GRAND PHILOSOPHE",
            _font(FONT_SERIF_BOLD, head_size), PALETTE["gold"])
    y += head_size + 26
    _ornament_row(draw, W / 2, y, inner_w * 0.6, PALETTE["gold"])
    y += 40

    # Corps : troncation élégante si la réflexion dépasse le format
    text = card.get("generated_text", "")
    limit = MAX_CHARS[size_key]
    if len(text) > limit:
        cut = text[:limit].rsplit(" ", 1)[0]
        text = cut.rstrip(".;") + "…"

    # Capitale ornée en début de réflexion
    body_size = int(34 * W / 1200) if size_key != "landscape" else 30
    cap_size = int(body_size * 2.6)
    body_font = _font(FONT_SERIF, body_size)
    cap_font = _font(FONT_SERIF_BOLD, cap_size)

    first = text[0]
    rest = text[1:]
    cap_w = draw.textlength(first, font=cap_font) + 14
    lines = _wrap(draw, rest, body_font, inner_w - cap_w)
    draw.text((margin + 30, y + cap_size // 6), first, font=cap_font,
              fill=PALETTE["gold_soft"])
    line_h = int(body_size * 1.55)
    tx = margin + 30 + cap_w
    avail = H - y - margin - (150 if card.get("attribution") else 90)
    max_lines = max(1, int(avail / line_h))
    for line in lines[:max_lines]:
        draw.text((tx, y), line, font=body_font, fill=PALETTE["ivory"])
        y += line_h

    # Attribution fictive (mode citation)
    if card.get("attribution"):
        y = H - margin - 150
        attr_font = _font(FONT_SERIF, int(body_size * 0.85))
        attr = card["attribution"]
        for line in _wrap(draw, attr, attr_font, inner_w)[:2]:
            w_ = draw.textlength(line, font=attr_font)
            draw.text((W - margin - 30 - w_, y), line,
                      font=attr_font, fill=PALETTE["muted"])
            y += int(body_size)
        # Badge : sécurité anti-confusion (§9.2)
        badge = "citation humoristique · attribution fictive"
        b_font = _font(FONT_SANS, int(body_size * 0.55))
        bw = draw.textlength(badge, font=b_font) + 28
        bx = W / 2 - bw / 2
        by = H - margin - 96
        draw.rounded_rectangle([bx, by, bx + bw, by + 30], radius=15,
                               outline=PALETTE["gold"], width=1)
        draw.text((bx + 14, by + 6), badge, font=b_font, fill=PALETTE["gold"])

    # pied de carte : sceau + URL du produit
    fy = H - margin - 42
    if not card.get("attribution"):
        _ornament_row(draw, W / 2, fy, inner_w * 0.5, PALETTE["gold"])
    foot_font = _font(FONT_SANS, max(13, int(15 * W / 1200)))
    _spaced(draw, W / 2, fy + 12, "LEGRANDPHILOSOPHE", foot_font, PALETTE["muted"], 4)

    return img
